// genetics-js/lib/crossover/base/OnePointCrossover.js
"use strict";
/*
 * @license
 * Copyright (c) 2019 Cristian Abrante. All rights reserved.
 * Licensed under the MIT License. See LICENSE in the project root for license information.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnePointCrossover = void 0;
const utils_1 = require("../../generator/utils");
const NPointsCrossover_1 = require("./NPointsCrossover");
class OnePointCrossover {
    cross(firstParent, secondParent, individualConstructor, engine = utils_1.Generator.DEFAULT_ENGINE) {
        return this.crossWith(firstParent, secondParent, { individualConstructor, engine });
    }
    crossWith(firstParent, secondParent, params) {
        const cross = new NPointsCrossover_1.NPointsCrossover();
        return cross.crossWith(firstParent, secondParent, {
            engine: params.engine,
            individualConstructor: params.individualConstructor,
            numberOfCrossoverPoints: 1,
        });
    }
}
exports.OnePointCrossover = OnePointCrossover;
